### ISSUE TITLE
API error 401 since yesterday

### ISSUE THREAD

**Issue by AdeTheux** [lukasroegner/homebridge-dyson-pure-cool#166]
Seems like authentication is failing again, I guess Dyson changed something again?

```
[3/3/2021, 8:55:16 AM] [DysonPureCoolPlatform] Checked user account.
[3/3/2021, 8:55:16 AM] [DysonPureCoolPlatform] Signing in.
[3/3/2021, 8:55:17 AM] [DysonPureCoolPlatform] Error while signing in. Status Code: 401
[3/3/2021, 8:55:17 AM] [DysonPureCoolPlatform] API could not be reached. Retry is disabled.
```

**Comment by jsalva23** [comment]
I'm having the exact same issue

3/3/2021, 9:18:16 AM [DysonPureCoolPlatform] Error while signing in. Status Code: 401
3/3/2021, 9:18:16 AM [DysonPureCoolPlatform] API could not be reached. Retry is disabled.

I'm using HOOBS 3.3.3, the version of the Dyson firmware is the latest: 1.8.7 all installed on a Raspberry Pi

**Comment by marhindev** [comment]
They really seem to have changed something. The release information for the latest update of the Dyson iOS app say something about new infterface for room climate devices (https://apps.apple.com/de/app/dyson-link/id993135524). It also states changes to authentification (2 step auth).

**Comment by henkied** [comment]
same here

[3/3/2021, 5:16:27 PM] [DysonPureCoolPlatform] Error while signing in. Status Code: 401
[3/3/2021, 5:16:27 PM] [DysonPureCoolPlatform] API could not be reached. Retry is disabled.

**Comment by jakemauer** [comment]
A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.

**Comment by AdeTheux** [comment]
> A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.

Indeed, works here as well. Thanks.

**Comment by marhindev** [comment]
If I have the Dyson app running on multiple devices, I assume that most likely I have to do this in all of them. Do you only run it on one device?

**Comment by bondskin** [comment]
This is what I get since yesterday:
[04/03/2021, 10:42:50] [DysonPureCoolPlatform] Error while signing in. Status Code: 429
[04/03/2021, 10:42:50] [DysonPureCoolPlatform] API could not be reached. Trying again soon...

**Comment by jsalva23** [comment]
> A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.

Not working for me... anyone else still having issues?

**Comment by bondskin** [comment]
> > A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.
> 
> Not working for me... anyone else still having issues?

did you wait a few minutes?

**Comment by jakemauer** [comment]
> If I have the Dyson app running on multiple devices, I assume that most likely I have to do this in all of them. Do you only run it on one device?
> […](#)
> Am 04.03.2021 um 04:18 schrieb Jake Mauer ***@***.***>: ﻿ A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge. — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

My assumption is logging in to one of the apps should do it. What happens when you try?

**Comment by jakemauer** [comment]
> > A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.
> 
> Not working for me... anyone else still having issues?

Are you on the latest app version? I had to supply a two-factor code sent to my email before I could login to the iOS app, which is new behavior I think.

**Comment by marhindev** [comment]
Logout and login, there providing the additional auth. code sent by mail from Dyson and then a reboot of Homebridge did it for me as well! It was sufficient to do it on one iOS device.

**Comment by AdeTheux** [comment]
Looks like the main requirements is re-authenticating via the official app, and getting that code via email. 
It's not really 2fa, but lols more like they're marking the email address as verified once this is done.

**Comment by jsalva23** [comment]
> > > A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.
> > 
> > 
> > Not working for me... anyone else still having issues?
> 
> did you wait a few minutes?

At what point did you waited? after login off on the mobile app? before login back in at the app? before login in on the plugin? how long did you waited?

**Comment by jsalva23** [comment]
> > > A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.
> > 
> > 
> > Not working for me... anyone else still having issues?
> 
> Are you on the latest app version? I had to supply a two-factor code sent to my email before I could login to the iOS app, which is new behavior I think.

yes, latest version, and yes I had to add the code sent to my email and yes that's new...

**Comment by jsalva23** [comment]
I actually got it working, for some reason the region acronym didn't match my region, I don't think I changed that recently I believe it was like that since I installed the plug-in for the first time. Not sure if the API was actually validating that, but it seems it is now. 

Anyways I changed to the right region and it's working now. Thank everyone for the help provided.

**Comment by robertburron** [comment]
I'm seeing the same thing. I've tried troubleshooting steps that others have recommended, but no luck. 
`[04/03/2021, 23:19:47] [DysonPureCoolPlatform] Signing in.
[04/03/2021, 23:19:47] [DysonPureCoolPlatform] Error while signing in. Status Code: 429
[04/03/2021, 23:19:47] [DysonPureCoolPlatform] API could not be reached. Trying again soon...
[04/03/2021, 23:20:47] [DysonPureCoolPlatform] Signing in.
[04/03/2021, 23:20:47] [DysonPureCoolPlatform] Error while signing in. Status Code: 429
[04/03/2021, 23:20:47] [DysonPureCoolPlatform] API could not be reached. Trying again soon...`

**Comment by AdeTheux** [comment]
> Looks like the main requirements is re-authenticating via the official app, and getting that code via email.
> It's not really 2fa, but lols more like they're marking the email address as verified once this is done.

Following the latest Dyson iOS app update from yesterday, this trick doesn't work anymore.

**Comment by marhindev** [comment]
So you are saying that it worked when you were on App version 5.0.21060, and after you installed 5.0.21061 it stopped working?

I am on 5.0.21060 and it still works so far.

**Comment by AdeTheux** [comment]
> So you are saying that it worked when you were on App version 5.0.21060, and after you installed 5.0.21061 it stopped working? I am on 5.0.21060 and it still works so far.
> […](#)

Not 100% sure but it was working fine until ~11pm yesterday. This morning the app was updated on my iPhone, and since then I got error 400 in HB again.

**Comment by Chewwy420** [comment]
I had same issue. Logged out of app and back in with new code. started working again...  Was doing work on my homebridge so multiple restarts and started doing it again... Logged out of app and back in with new code and it is working again.. Seems there is a timeout of sorts.

**Comment by jsalva23** [comment]
Mine is working fine since yesterday, when I fixed the region code I was already in the latest version of the app

**Comment by marhindev** [comment]
I updated now and it still seems to be working.

**Comment by philipamour** [comment]
If you can't sign in or sign up, this is what I got from Dyson Support after spending hours on the phone and web chat:

```We’re sorry to see you’ve been unable to access your Dyson account with us.We are aware of a technical issue with the accounts section which stops a small percentage of our owners from accessing their accounts or changing their password, our IT department are aware of the issue and this is currently a work in progress. We can’t advise when this issue will be fully resolved, sadly, due to the current situation and increased workloads for our IT department this may take some months.

If you’d like to place an order, we’d suggest checking out as a guest for the time being, alternatively you’re certainly welcome to create a new online Dyson account with a different email address.```

**Comment by AdeTheux** [comment]
> ```
> 
> If you’d like to place an order, we’d suggest checking out as a guest for the time being, alternatively you’re certainly welcome to create a new online Dyson account with a different email address.```
> ```
Not really related to this issue.

**Comment by philipamour** [comment]
Asked them if this is relevant to Dyson Link, they said yes.

**Comment by philipamour** [comment]
I've had to log out of Dyson link, create a new account with a new email, re-pair the device and then I was able to use my new credentials in this plugin config. All works now.

**Comment by Stivenskyrah** [comment]
> A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.

Worked for me too thanks!

**Comment by ronnieliew** [comment]
Logging into the app (and entering the 6-digit “2FA”) worked for me as well.
I was also prompted to accept the new terms of service but I’m not sure if this is related or that it’s been a while since I logged into the app.

**Comment by robertburron** [comment]
I was able to resolve my issue by logging out of the app, deleting the app, reinstalling it, and going through the login process again. Just logging out and back in did not trigger the terms of use page, which obviously plays a part in fully authenticating the account. The "2FA" portion went smoothly. After gaining access to the app, I restarted homebridge and everything registered. I made sure to copy my device credentials and add them to the config this time around so I'm not left hanging in the future.

**Comment by AdeTheux** [comment]
It's frustrating how complex this is. I guess there's a specific sequence to follow carefully, which is why it works for some and not for others (logging out worked the 1st time for me, not anymore, even after accepting new T&Cs).
Just thought of the fact the same account is used on the Dyson app on my iPad, I haven't used the app there in months, so I'll try to sign out and accept T&Cs as well 🤞🏼

**Comment by jsalva23** [comment]
Hey @lukasroegner not sure if this issue can get fixed at a plug-in level, can we get your comments on what's going on?

**Comment by catsartcox** [comment]
I have tried logging in and out of the app and even uninstalling it and reloading it ( after that step I got the ToS to accept ), but I cannot get it to log in. Just am getting status code 429.

**Comment by lukasroegner** [comment]
@jsalva23 I cannot think of a proper solution, yet. 

I want to wait some time before making huge changes to the plugin, as it seems that Dyson has been constantly changing the login flow. For now, I can only recommend to use the "store credentials" option so that the Dyson API is no longer needed.

**Comment by AdeTheux** [comment]
> I can only recommend to use the "store credentials" option so that the Dyson API is no longer needed.

@lukasroegner is there any way to get those credentials if I can't currently log in? Like curl or wireshark? Thanks

**Comment by twentworth12** [comment]
Like @robertburron I had to uninstall the iOS app to trigger the trigger the terms of service checkbox. Working again now.

**Comment by AdeTheux** [comment]
Just doing it on iPhone was not sufficient.
For those of you who have installed the app on multiple devices: I deleted it on all devices, reinstalled, and accepted T&Cs on all. After a HB reboot, it's all back on 🙂

**Comment by bondskin** [comment]
It was working fine for 1-2 days here, now the error 401 / 429 is back. logging off and on did not help this time.
I think I'll disable the plugin for the time being.

**Comment by Andybishop** [comment]
I had the same, logged out of the app and logged back in, entered code sent by email, restarted Hoobs service and its now working fine.

**Comment by fcmarriott** [comment]
> > > A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.
> > 
> > 
> > Not working for me... anyone else still having issues?
> 
> Are you on the latest app version? I had to supply a two-factor code sent to my email before I could login to the iOS app, which is new behavior I think.

This workaround worked for me. Logged out, logged back in, with the 6-digit code sent in email. Took ~45 seconds of unsuccessful login to finally work. iOS app version 5.0.21061.

**Comment by marhindev** [comment]
> 
> 
> It was working fine for 1-2 days here, now the error 401 / 429 is back. logging off and on did not help this time.
> I think I'll disable the plugin for the time being.

Same with me. It came back one time, and a reboot of homebridge solved it. Came back now again, and reboot did not work.
So I am currently using the credentials in the config, which works fine.

**Comment by mwhaley99** [comment]
How do I get/view the logs via the Hoobs box in order to get the credentials?

**Comment by marhindev** [comment]
HOOBS has a built in log viewer in the UI.

**Comment by mwhaley99** [comment]
I looked in there and it does not show any password information.

**Comment by marhindev** [comment]
You need to search for an old entry where the plug-in could still successfully login. The credentials are printed there.

**Comment by Krillle** [comment]
To find credentials from the times when the plug-in could still successfully login: Where to find older log files in the homebridge.io standard image setup? 
/var/lib/homebridge/homebridge.log contains only the very recent entries.

**Comment by fcmarriott** [comment]
I followed the instructions above to make it work successfully, then copied the creds into the config. Working great now.

**Comment by Krillle** [comment]
The above described process worked for me as well, however it was not working while iPhone was in another WiFi and Homebridge accessed via Zerotier VPN. 

- **iPhone and Homebridge have to be in the same network when logging in to Dyson API, probably need to have the same outside IP address.**
- Logout from Dyson API in Dyson App
- Login to Dyson API in Dyson App (un-/reinstalling app or to reapproving terms of service (ToS) was not necessary) 
- Login to Dyson API in Homebridge via restarting Homebridge or in my case schedule retry 

Devices are back online. Copy credentials to the plugin settings to be safe for the next restart.

**Comment by bondskin** [comment]
I found the credentials in the homebrige.log and used it in the Homebridge config, this made it work for me

**Comment by arkku** [comment]
Unfortunately still doesn't work for me even with the above (i.e., sign out of mobile app, login again on same network as the Homebridge, restart Homebridge). There's a sort of 2FA in the app (6-digit code to email), I'm guessing that is the problem.

**Comment by rohrkemper** [comment]
> I found the credentials in the homebrige.log and used it in the Homebridge config, this made it work for me

Can you explain where to find the credentials in the log file. I see some text saying that they will be upcoming in the log. But I don't see the actual credentials.

**Comment by bondskin** [comment]
> > I found the credentials in the homebrige.log and used it in the Homebridge config, this made it work for me
> 
> Can you explain where to find the credentials in the log file. I see some text saying that they will be upcoming in the log. But I don't see the actual credentials.

@rohrkemper you need to go back in the log until the last time the plugin connected successfully, there it prints the credentials. if you cycle your logs or don't archive it, you will not see this information until you are able to login again
It looks something like:
"Credentials for device with serial number NN2-EU-YXZ are: eyJTZXJpYW..."

**Comment by rohrkemper** [comment]
Whoops I already deleted all logs when I reinstalled.  Other ideas?

**Comment by sweetw0r** [comment]
Decided to give it a go:
https://mega.nz/folder/ULwAzD6K#twtjwaSJWEF1AQrsdFM_sQ

There are a postman collection and `Dyson.apk` source.

00. Status
01. Get a code by `EMAIL_PWD_2FA`
02. Credentials. It is incomplete. See if you can figure this one out.

**Comment by tkaudioservices** [comment]
@sweetw0r Did you manage to get this to work? Is this a way of extracting your machines credentials, I didn't know mine before the 2FA .

**Comment by siim911** [comment]
I USE TP06
and i have the same issue i tried to get the credentials but i failed. this is happened after the new sign in changes which require to enter a code sent to the email

[DysonPureCoolPlatform] Device credentials not stored, asking Dyson API. If you want to prevent communication with the Dyson API, copy the credentials for each device from the coming log entries to the config.json.
[3/31/2021, 7:07:16 PM] [DysonPureCoolPlatform] Checked user account.
[3/31/2021, 7:07:16 PM] [DysonPureCoolPlatform] Signing in.
[3/31/2021, 7:07:22 PM] [DysonPureCoolPlatform] Error while signing in. Status Code: 429

**Comment by arkku** [comment]
I got the same error, but finally managed to get it working like this:

1. stop Homebridge
2. log out of Dyson from their mobile app
3. type in the command to restart Homebridge but don't press enter
4. log in to Dyson from their mobile app (on the same network as your Homebridge) and complete the 2FA
5. immediately as you log in, press enter on the Homebridge restart command

**Comment by tkaudioservices** [comment]
Is there no other way to retrieve your credentials, tried time and time again to time to mobile app login and Homebridge restart; to no avail

**Comment by Krillle** [comment]
@tkaudioservices Make sure your phone and homebridge are in the same network, so presenting same outer IP to Dyson API. Like that (and only like that) it worked for me. 
(I didn’t even need to hurry as @arkku suggests.)

**Comment by arkku** [comment]
Could be that it's simply unrealiable and not actually as time-sensitive as I thought, i.e., retry a few times, and also wait some time between retries since I think it blocks you for too many failed attempts.

**Comment by sweetw0r** [comment]
Still need to figure out that final call to receive credentials. I suspect it is to `/v3/userregistration/email/verify` where you are supposed to pass `"otpCode"` that gets sent to your email after calling the `/v3/userregistration/mobile/auth`.

Can somebody with Java knowledge look at this syntax to figure out the request scheme?

sources/p033c3/C1545d.java
```
public C1545d(C3632i iVar, C8058u uVar, String str, String str2, String str3, String str4, String str5) {
        C9482o.m34106c(iVar, "httpRequestFactory");
        C9482o.m34106c(uVar, "userManagementEndpointsConfig");
        C9482o.m34106c(str, "countryCode");
        C9482o.m34106c(str2, SessionInfoKeys.Email);
        C9482o.m34106c(str3, "password");
        C9482o.m34106c(str4, "challengeId");
        C9482o.m34106c(str5, "otpCode");
        this.f4706b = iVar;
        this.f4707c = uVar;
        this.f4708d = str;
        this.f4705a = C10600m.m38108e(C15455l0.m55355l(C14509u.m52266a(SessionInfoKeys.Email, str2), C14509u.m52266a("password", str3), C14509u.m52266a("challengeId", str4), C14509u.m52266a("otpCode", str5)));
```

**Comment by jonathanswenson** [comment]
Looks like the secret here is the wifipassword of the device printed on the manual / sticker on the front of the fan. 

It is first hashed (SHA512) and then base64 encoded. that is the “password” that you can also fetch with the manifests: 
```
hash = crypto.createHash('sha512')
hash.update(“<wifi_password_here>”)
buffer = hash.digest()
buffer.toString(“base64”)
```

However, this alone is not sufficient as the value for the “credentials” field for the plugin. 

Looking at the source it is the base64 encoded JSON stringified apiConfig (as fetched from the manifest endpoint) with the addition of a “password” field to represent the sha512 hashed / base 64 encoded json string. 

see (here)[https://github.com/lukasroegner/homebridge-dyson-pure-cool/blob/7287ae32d1e947242ab5c7655a79d3adbe3a2bf6/src/dyson-pure-cool-platform.js#L240] and (here)[https://github.com/lukasroegner/homebridge-dyson-pure-cool/blob/7287ae32d1e947242ab5c7655a79d3adbe3a2bf6/src/dyson-pure-cool-platform.js#L268]

For me this was: 
```
{"Serial":"<serial>”,"Name":"<device_name>“,"Version":"<device_version>“,"LocalCredentials":"<from_manifest>“,"AutoUpdate":true,"NewVersionAvailable":true,"ProductType":"475","ConnectionType":"wss", "password": "<hashed/base64encoded wifi password>”}
```

It is possible to log in using the new authentication flow (that requires the 2FA email). I followed [this comment](https://github.com/etheralm/libpurecool/issues/37#issuecomment-791635081) to figure out how to do that. 

its very possible that one can circumvent having to authenticate with dyson completely if you have the wifi password. It appears that the plugin is only using part of this: https://github.com/lukasroegner/homebridge-dyson-pure-cool/blob/7287ae32d1e947242ab5c7655a79d3adbe3a2bf6/src/dyson-pure-cool-platform.js#L323

in particular the fields `apiConfig.Name, apiConfig.Serial, apiConfig.ProductType, apiConfig.Version, apiConfig.password`

was able to do the following given that I know the serial (from sticker or dyson link), name (from dyson link), version (don’t know how to get this -- not sure if it matters), product type (from the front sticker final 3 digits of product SSID) and wifi password. 

```
wifiPassword = “<wifi_password>”
hash = crypto.createHash('sha512')
hash.update(wifiPassword)
buffer = hash.digest()
password =  buffer.toString(“base64”)

serial = “<serial>”
name = “<name>”
version = “<version>”
productType = “<productType>”

json = {"Serial": serial, “Name": name,"Version": version,"ProductType":productType, “password": password}

credentials = Buffer.from(JSON.stringify(json)).toString('base64')
```

Hopefully that helps someone who has been struggling to the re-auth flow to work.

**Comment by coraxx** [comment]
@sweetw0r  I just quickly played around with the new API calls a bit (in python). This is how you could login with the 2FA and receive a token. Though I do not know how new API calls look like using this token.

If someone is interested, here some pseudo code (working but no logic combining them):

```python
import requests

# This checks if the user/email exists
url = 'https://appapi.cp.dyson.com/v3/userregistration/email/userstatus'
params = {'country': 'DE'}
payload = {'Email': 'XYZ@domain.tld'}
headers = {'user-agent': 'android client'}
r = requests.post(url, params=params, headers=headers, json=payload, verify = False)
# returns something like:
# {"accountStatus":"ACTIVE","authenticationMethod":"EMAIL_PWD_2FA"}
print (r.text)


# This will trigger sending the email with 2FA code
url = 'https://appapi.cp.dyson.com/v3/userregistration/email/auth'
params = {'country': 'DE'}
payload = {'Email': 'XYZ@domain.tld', 'Password': 'SecretDysonAccountPassword'}
headers = {'user-agent': 'android client'}
r = requests.post(url, params=params, headers=headers, json=payload, verify = False)
# returns something like:
# {"challengeId":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"}
print (r.text)


# Get the challenge id from previous call and the 6 digit code from the 2FA email
url = 'https://appapi.cp.dyson.com/v3/userregistration/email/verify'
params = {'country': 'DE'}
payload = {'Email': 'XYZ@domain.tld', 'Password': 'SecretDysonAccountPassword', "challengeId":"result from previous call", 'otpCode': '6 digit code received via email'}
headers = {'user-agent': 'android client'}
r = requests.post(url, params=params, headers=headers, json=payload, verify = False)
# returns something like:
# {"account":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx","token":"xxxxx-xxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxxxxxxx","tokenType":"Bearer"}
print (r.text)```

**Comment by coraxx** [comment]
Small update: after getting the token, all old API calls work as expected. Authentication is changed from "Basic" to "Bearer":

```python
url = 'https://appapi.cp.dyson.com/v2/provisioningservice/manifest'
params = {'country': 'DE'}
# token id which gets returned after sending the 2FA code in the form of xxxxx-xxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxxxxxxx
headers = {'user-agent': 'android client', 'Authorization': 'Bearer xxxxx-xxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxxxxxxx'}
r = requests.get(url, params=params, headers=headers, verify = False)
# returns all your dyson products
print (r.text)
```

**Comment by tkaudioservices** [comment]
> Small update: after getting the token, all old API calls work as expected. Authentication is changed from "Basic" to "Bearer":
> 
> ```python
> url = 'https://appapi.cp.dyson.com/v2/provisioningservice/manifest'
> params = {'country': 'DE'}
> # token id which gets returned after sending the 2FA code in the form of xxxxx-xxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxxxxxxx
> headers = {'user-agent': 'android client', 'Authorization': 'Bearer xxxxx-xxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxxxxxxx'}
> r = requests.get(url, params=params, headers=headers, verify = False)
> # returns all your dyson products
> print (r.text)
> ```

Has anyone successfully pieced this together to receive Credentials?

**Comment by coraxx** [comment]
Yeah, it's very much just a proof of concept. But should be easy enough to update the plugin here.
Biggest issue I see, is to make entering the 2FA code easy enough for the end user.

Sadly my JavaScript knowledge is very limited, otherwise I would offer a pull request.
If I have time tomorrow I will give it a try.

**Comment by jonathanswenson** [comment]
@tkaudioservices I have successfully gone though the new authorization flow (V3) using curl and used the final returned bearer token to fetch the “password”. That password can be generated without having to do the auth flow if you have your wifi password (as printed on the sticker on the front / in the manual). 

Unfortunately, that is not sufficient to generate the “credentials” as required by the plugin though. You’ll have to encode that password (and a few other bits of information properly) to generate one of these.

**Comment by sweetw0r** [comment]
Thanks @coraxx Just updated my original https://github.com/lukasroegner/homebridge-dyson-pure-cool/issues/166#issuecomment-808862482

Here is an example of response for `v3/userregistration/email/verify` request. _Token_ is the credentials:
```
{
    "account":"72b2af5b-2ed0-4c41-b779-2cd1a7082788",
    "token":"1RyaUBeez-uLXQ71jckTcxWmc4TsQE5BPVc8I6F0cGw",
    "tokenType":"Bearer"
}
```

**Comment by porowskid** [comment]
I suspect there's something else going on as well.  I've got credentials for my devices stored (to avoid the API) and even though the manifest localCredentials (gotten with the 2fa/token process above) still match the creds I had stored, they no longer seem to base64 decode as they did.  I haven't dug too far in yet, though; I thought it was an issue of attaching the bearer token to the device comms as well as to the API, but the plugin is failing at what appears to be the config-credential decode step.

**Comment by dampney** [comment]
Signing into the Dyson app again and Homebridge at the same time didn't work. 

I've sent a negative survey to Dyson over the new app.

Hopefully we can find a work around with this change Dyson did to break the API.

**Comment by coraxx** [comment]
I think I have a version ready for the plugin. Only "problem" is, that it is not a nice user experience ^^

First you have to start the homebridge to retrieve the challengeID in the log. Copy that into the config, as well as the one time password code you get via email.
Then restart homebridge and get the token in the log, copy that into the plugin config aswell.

Then it should work fine from there on. I just tested it and I can still logout and log back in with 2FA on my smartphone.

Only problem is, when you have too many API calls in a short time. Dyson API has quite the hard API rate limiting.

**Comment by wbsmith2** [comment]
> > A workaround that just worked for me was to log out of the latest iOS app, log back in, then reboot homebridge.
> 
> Hi @jakemauer, I tried that but no luck
> 
> EDIT: correction, logging out/in via iOS app, restarting Homebridge fixes it for me. It just takes a few minutes

This fix worked for me

**Comment by raiubreaksthings** [comment]
Workaround, 

Log out of the app on the phone. Then restart homebridge. Let homebridge login. Then log back into the app. I am good to go now.

**Comment by Soldiiier** [comment]
I think the log out / log back in steps are overkill.
All you need to do is enter the Dyson app and then wait for the plugin to try its next login. mine was set to 30 minutes, so too long to have to wait. though this doest resolve the issue entirely. it still comes back about once a week. 

I've now gone to the 'credentials' method.

**Comment by sweetw0r** [comment]
Update [Dyson API.postman_collection.json](https://mega.nz/folder/ULwAzD6K#twtjwaSJWEF1AQrsdFM_sQ) with `03.  Get Credentials` requests to retrieve the credentials:

```
[
    {
        "Serial":"C4B-US-XXXXXXX",
        "Name":"Bedroom",
        "Version":"ECG2PF.46.00.007.0003",
        "LocalCredentials":"XXXXXXXXXXXXXXXXXXXXXXXXXXX/XXXXXXXXXXXXXXXXXXXXXXXXXXX/XXXXXXXXXXXXXXXXXXXXXXXXXXX",
        "AutoUpdate":false,
        "NewVersionAvailable":true,
        "ProductType":"358",
        "ConnectionType":"wss"
    }
]
```

**Comment by lukasroegner** [comment]
I finally had some time to find a viable solution for 2FA. With the help of @sweetw0r and @coraxx I added a separate flow for retrieving the credentials for all your devices. The plugin now has a small website where you can retrieve the credentials by logging into the Dyson account. The plugin no longer directly communicated with the Dyson API. New version 2.0.0 is live on npm.

Let me know that you think. I haven't tried the non-2FA flow, as my account has already been migrated to 2FA, however, the old flow is also implemented in the "website" for retrieving credentials.

**Comment by iMonZ** [comment]
None of these approaches helped.
When I visit the Authentication website on port 48000 I get directly these errors in the console:

> [Error] TypeError: undefined is not an object (evaluating 'crypto.subtle.generateKey')
> 	createKey (topee-content.js:4028)
> 	TextCrypto (topee-content.js:3981)
> 	(anonymous function) (topee-content.js:3869)
> 	__webpack_require__ (topee-content.js:20)
> 	(anonymous function) (topee-content.js:3312)
> 	__webpack_require__ (topee-content.js:20)
> 	(anonymous function) (topee-content.js:3292)
> 	__webpack_require__ (topee-content.js:20)
> 	(anonymous function) (topee-content.js:3279)
> 	__webpack_require__ (topee-content.js:20)
> 	(anonymous function) (topee-content.js:123)
> 	(anonymous function) (topee-content.js:229)
> 	__webpack_require__ (topee-content.js:20)
> 	(anonymous function) (topee-content.js:84)
> 	Global Code (topee-content.js:85)
> [Error] ReferenceError: Can't find variable: chrome
> 	(anonymous function) (content.js:21826)
> 	injectContent (content.js:21839)
> 	injectContentOnce (content.js:21)
> [Error] ReferenceError: Can't find variable: chrome
> 	injectContent (styles.js:3253)
> 	injectContentOnce (styles.js:3249)
> [Error] ReferenceError: Can't find variable: chrome
> 	injectContent (fonts.js:391)
> 	injectContentOnce (fonts.js:387)
> [Error] Unhandled Promise Rejection: null
> 	promiseEmptyOnRejected
> 	promiseReactionJob
> [Error] Unhandled Promise Rejection: null
> 	(anonymous function)
> 	rejectPromise
> 	promiseReactionJob
